internal/typeinfo/typeinfo.go

```diff
@@ -0,0 +1,115 @@
+package typeinfo
+
+import (
+	"fmt"
+	"reflect"
+	"regexp"
+	"strings"
+	"sync"
+)
+
+var cacheMutex sync.RWMutex
+var cache = make(map[reflect.Type]*Info)
+
+// Reflect will return the Info of a given type,
+// generating and caching as required.
+func TypeInfo(value any) (*Info, error) {
+	if value == (any)(nil) {
+		return &Info{}, fmt.Errorf("cannot reflect nil value")
+	}
+
+	v := reflect.ValueOf(value)
+	v = reflect.Indirect(v)
+
+	cacheMutex.RLock()
+	info, found := cache[v.Type()]
+	cacheMutex.RUnlock()
+	if found {
+		return info, nil
+	}
+
+	info, err := generate(v)
+	if err != nil {
+		return &Info{}, err
+	}
+
+	cacheMutex.Lock()
+	cache[v.Type()] = info
+	cacheMutex.Unlock()
+
+	return info, nil
+}
+
+// generate produces and returns reflection information for the input
+// reflect.Value that is specifically required for SQLAir operation.
+func generate(value reflect.Value) (*Info, error) {
+	// Dereference the value if it is a pointer.
+	value = reflect.Indirect(value)
```

Comment:
This was already done by the function that calls this function. In go-yaml, we don't do either of those because the method expects an actual struct, since there's already handling for type-related information before reaching the point where we decide to call this function, and as we covered a few times, this is only meant to be used in structs, so whatever is calling it needs to know somehow.  It's okay to use a different convention, of course, but as it is right now it feels like there's no clear behavior defined, and we just call it everywhere just in case. What's the overall idea?